Include folders in the result of getAdapterObjects

AlCalzone opened the pull request:
fixes: #1085 

I have a bunch of questions when looking at this code - lets discuss

AlCalzone commented on lib/adapter.js:
```diff
@@ -1887,28 +1887,40 @@ function Adapter(options) {
+
+                        adapterObjects.getObjectView('system', 'folder', {startkey: this.namespace + '.', endkey: this.namespace + '.\u9999', include_docs: true}, (err, _folders) => {
+                            if (_channels && _channels.rows) {
+                                for (let c = _channels.rows.length - 1; c >= 0; c--) {
```
Why are these objects assigned in the reverse order?

AlCalzone commented on lib/adapter.js:
```diff
@@ -1887,28 +1887,40 @@ function Adapter(options) {
-                                this.oStates = {};
+                            if (_devices && _devices.rows) {
+                                for (let d = _devices.rows.length - 1; d >= 0; d--) {
+                                    objects[_devices.rows[d].id] = _devices.rows[d].value;
```
Why are channels added before devices?

AlCalzone commented on lib/adapter.js:
```diff
@@ -1887,28 +1887,40 @@ function Adapter(options) {
-                        return tools.maybeCallback(callback, objects);
+                            if (_states && _states.rows) {
+                                if (options.states) {
+                                    this.oStates = {};
```
What is this (and the lines below about)? Why are we nulling the states cache when reading all adapter objects?

AlCalzone commented on lib/adapter.js:
```diff
@@ -1887,28 +1887,40 @@ function Adapter(options) {
-                            for (let c = _channels.rows.length - 1; c >= 0; c--) {
-                                objects[_channels.rows[c].id] = _channels.rows[c].value;
-                            }
+
```
Should we bail early with an error when one of these calls returns an error?

AlCalzone replied:
Hm that comment wasn't meant to go here... I mean the `err` in the getObjectView callback

AlCalzone commented:
And most of all: why is this very specific behavior not commented? 🤐

foxriver76 commented:
Lgtm but would be great to implement new code in async style we have `objects.getObjectViewAsync` too.  Regarding your questions we most likely need info from  @GermanBluefox

AlCalzone commented:
I've changed it as suggested and I've removed the back-to-front assignment since I don't see why it should be necessary.
Clearing oStates needs to be discussed IMO

AlCalzone commented:
And the test failures **seem** unrelated but I'm not sure.

foxriver76 commented:
First test failing at https://github.com/ioBroker/ioBroker.js-controller/blob/36549540d4a51bc7212bc2f5188de12a51a6d264/test/lib/testObjectsFunctions.js#L111 is susp

AlCalzone commented:
I'll check it later, thanks

AlCalzone commented:
Mehhh:
The error is: `Cannot find search "folder" in "system"`. Is that something bugged in the tests or do we need to add a method for that to the DB?

foxriver76 commented:
I've made a PR against your branch to fix it. It's only a testing issue, the view is already existing in production io-pack https://github.com/ioBroker/ioBroker.js-controller/blob/36549540d4a51bc7212bc2f5188de12a51a6d264/io-package.json#L276

GermanBluefox commented:
Thank you!
